packages/core/src/external/carequality/iti55/process-incoming-iti-55.ts

```diff
@@ -0,0 +1,205 @@
+  InternalError,
+  PatientAddressRequestedError,
+  LivingSubjectAdministrativeGenderRequestedError,
+} from "./validating-iti55";
+import { matchPatients, jaroWinklerSimilarity } from "../../mpi/match-patients";
+import { normalizePatientDataMPI } from "../../mpi/normalize-patient";
+import { mergePatients, mergeWithFirstPatient } from "../../mpi/merge-patients";
+import { capture } from "../../../util/notifications";
+import { getEnvVarOrFail } from "../../../util/env-var";
+
+import axios from "axios";
+const ossApi = axios.create();
+const SIMILARITY_THRESHOLD = 0.96;
+const METRIPORT_HOME_COMMUNITY_ID = "urn:oid:2.16.840.1.113883.3.9621";
+
+type MPIBlockParams = {
+  dob: string;
+  genderAtBirth: string;
+  firstNameInitial?: string;
+  lastNameInitial?: string;
+};
+
+function constructErrorResponse(
+  payload: PatientDiscoveryRequestIncoming,
+  codingSystem: string,
+  code: string,
+  error: string,
+  patientMatch: boolean | null = null
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: patientMatch,
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+    operationOutcome: {
+      resourceType: "OperationOutcome",
+      id: payload.id,
+      issue: [
+        {
+          severity: "error",
+          code: "processing",
+          details: {
+            coding: [{ system: codingSystem, code: code }],
+            text: error,
+          },
+        },
+      ],
+    },
+  };
+}
+
+function constructNoMatchResponse(
+  payload: PatientDiscoveryRequestIncoming
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: false,
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+  };
+}
+
+function constructMatchResponse(
+  payload: PatientDiscoveryRequestIncoming,
+  patient: PatientDataMPI
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: true,
+    xcpdPatientId: {
+      id: patient.id,
+      system: "000", // TBD	- what is this
+    },
+    patientResource: convertPatientToFHIR(patient),
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+  };
+}
+
+export async function processIncomingRequest(
+  payload: PatientDiscoveryRequestIncoming
+): Promise<PatientDiscoveryResponseOutgoing> {
+  const apiClient = apiClientMPIBlockEndpoint();
+  console.log("payload", payload);
+
+  let patient: PatientDataMPI;
+  try {
+    patient = validateFHIRAndExtractPatient(payload.patientResource);
+    console.log("patient", patient);
+  } catch (error) {
+    if (error instanceof InternalError) {
+      return constructErrorResponse(
+        payload,
+        "1.3.6.1.4.1.19376.1.2.27.3",
+        "InternalError",
+        error.message,
+        null
```

Comment:
we normally try to use undefined and not null, but iti55 specs require null if there is an internal server error